In [40]:
import pandas as pd
import numpy as np

In [41]:
#Q1
pd.__version__

'3.0.5'

Q1. What version of Pandas did you install? '3.0.5'


In [42]:
data = pd.read_csv('car_fuel_efficiency_2026.csv')

In [43]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   model_year           10000 non-null  int64  
 1   origin               10000 non-null  str    
 2   fuel_type            10000 non-null  str    
 3   drivetrain           10000 non-null  str    
 4   num_doors            10000 non-null  int64  
 5   engine_displacement  10000 non-null  int64  
 6   num_cylinders        10000 non-null  int64  
 7   horsepower           9123 non-null   float64
 8   vehicle_weight       10000 non-null  int64  
 9   acceleration         9736 non-null   float64
 10  fuel_efficiency_mpg  10000 non-null  float64
dtypes: float64(3), int64(5), str(3)
memory usage: 859.5 KB


Q2. Records count
How many records are in the dataset? 10000

In [44]:
#Q1 
data['fuel_type'].unique()

<StringArray>
['Gasoline', 'Diesel', 'Hybrid']
Length: 3, dtype: str

Q3. Fuel types
How many fuel types are presented in the dataset? 3,['Gasoline', 'Diesel', 'Hybrid']

In [45]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   model_year           10000 non-null  int64  
 1   origin               10000 non-null  str    
 2   fuel_type            10000 non-null  str    
 3   drivetrain           10000 non-null  str    
 4   num_doors            10000 non-null  int64  
 5   engine_displacement  10000 non-null  int64  
 6   num_cylinders        10000 non-null  int64  
 7   horsepower           9123 non-null   float64
 8   vehicle_weight       10000 non-null  int64  
 9   acceleration         9736 non-null   float64
 10  fuel_efficiency_mpg  10000 non-null  float64
dtypes: float64(3), int64(5), str(3)
memory usage: 859.5 KB


Q4. Missing values
How many columns in the dataset have missing values? 2 columns

In [46]:
data.loc[data["origin"]=="Asia"]["fuel_efficiency_mpg"].max()

np.float64(41.2)

Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia? 41.2

In [47]:
data["horsepower"].median()

np.float64(254.0)

In [48]:
data["horsepower"].mode()

0    252.0
Name: horsepower, dtype: float64

In [49]:
data["horsepower"] =  data["horsepower"].fillna(value = 252)

In [50]:
data["horsepower"].median()

np.float64(252.0)

Q6. Median value of horsepower
Find the median value of the horsepower column in the dataset.
Next, calculate the most frequent value of the same horsepower column.
Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
Now, calculate the median value of horsepower once again.
Has it changed?

Yes, it decreased from 254 to 252

In [51]:
data_asia = data.loc[data["origin"]=="Asia"]

In [52]:
X = np.array(data_asia[["vehicle_weight","model_year"]][:7])

In [53]:
#Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
XTX = X.T @ X

In [55]:
#Invert XTX.
XTX_inv = np.linalg.inv(XTX)

In [ ]:
#Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
Y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [58]:
#step1 = Multiply the inverse of XTX with the transpose of X, 

step1 = XTX_inv @ X.T

#step2 = and then multiply the result by y. Call the result w.

step2 = step1 @ Y

In [61]:
#What's the sum of all the elements of the result?
np.sum(step2)

np.float64(0.36919696904925486)

Q7. Sum of weights
Select all the cars from Asia
Select only columns vehicle_weight and model_year
Select the first 7 values
Get the underlying NumPy array. Let's call it X.
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
Invert XTX.
Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
What's the sum of all the elements of the result?

answer: 0.369